sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/TimeoutHandler.java

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.netty.implementation;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.EmptyByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * This class observes a network calls request and response and manages triggering a timeout if the operation stalls.
+ * <p>
+ * Timeout tracking is split into three distinct categories of write, response, and read timeout. All timeout operations
+ * use a minimum timeout period of 1 millisecond.
+ * <p>
+ * Write timeouts track data being sent through the network sockets. Queueing and callbacks are used to ensure the write
+ * operation isn't being stalled or timing out. Each time a write operation is processed in this handler an event may be
+ * added to the tracking queue depending on how quickly the data is sent through the network socket. If the write
+ * doesn't complete instantaneously a write callback will be added into the tracking queue, if this completes before the
+ * timeout period it will be cancelled and removed. If the write operation progresses too slowly a {@link
+ * TimeoutException} will be thrown and all other write operations being tracked in the queue will be cancelled.
+ * <p>
+ * Response timeout tracks whether the server responds in a timely manner once the request has completed sending. This
+ * timeout begins once all write operations have processed.
+ * <p>
+ * Read timeout tracks that network response read operations are periodically happening. The first channel read
+ * completion will cancel the response timeout and begin read timeout tracking. Each time a new read completes the
+ * timeout resets, this is different than write timeouts which have individual timeouts for each write operation. The
+ * read timeout is removed when the channel handler is removed or when the channel becomes inactive after the full
+ * response is consumed.
+ */
+public final class TimeoutHandler extends ChannelDuplexHandler {
```

Comment:
Is there a way to unit test this? Doesn't netty's client builder provide options to set the timeouts? Should this be a change that should be included in core netty library itself?
I think we'll miss out on any improvements Netty makes to this `ChannelDuplexHandler` by overriding some of these methods. If they were to own this and provides hooks for it, then it would be ideal.